precompiles/balances-erc20/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+
+/// Associates pallet Instance to a prefix used for the Approves storage.
+/// This trait is implemented for () and the 16 substrate Instance.
+pub trait InstanceToPrefix {
+	/// Prefix used for the Approves storage.
+	type ApprovesPrefix: StorageInstance;
+}
+
+// We use a macro to implement the trait for () and the 16 substrate Instance.
+macro_rules! impl_prefix {
+	($prefix:ident, $instance:ty, $name:literal) => {
+		pub struct $prefix;
+
+		impl StorageInstance for $prefix {
+			const STORAGE_PREFIX: &'static str = "Approves";
+
+			fn pallet_prefix() -> &'static str {
+				$name
+			}
+		}
+
+		impl InstanceToPrefix for $instance {
+			type ApprovesPrefix = $prefix;
+		}
+	};
+}
+
+impl_prefix!(ApprovesPrefix0, (), "Erc20Instance0Balances");
+impl_prefix!(ApprovesPrefix1, Instance1, "Erc20Instance1Balances");
+impl_prefix!(ApprovesPrefix2, Instance2, "Erc20Instance2Balances");
+impl_prefix!(ApprovesPrefix3, Instance3, "Erc20Instance3Balances");
+impl_prefix!(ApprovesPrefix4, Instance4, "Erc20Instance4Balances");
+impl_prefix!(ApprovesPrefix5, Instance5, "Erc20Instance5Balances");
+impl_prefix!(ApprovesPrefix6, Instance6, "Erc20Instance6Balances");
+impl_prefix!(ApprovesPrefix7, Instance7, "Erc20Instance7Balances");
+impl_prefix!(ApprovesPrefix8, Instance8, "Erc20Instance8Balances");
+impl_prefix!(ApprovesPrefix9, Instance9, "Erc20Instance9Balances");
+impl_prefix!(ApprovesPrefix10, Instance10, "Erc20Instance10Balances");
+impl_prefix!(ApprovesPrefix11, Instance11, "Erc20Instance11Balances");
+impl_prefix!(ApprovesPrefix12, Instance12, "Erc20Instance12Balances");
+impl_prefix!(ApprovesPrefix13, Instance13, "Erc20Instance13Balances");
+impl_prefix!(ApprovesPrefix14, Instance14, "Erc20Instance14Balances");
+impl_prefix!(ApprovesPrefix15, Instance15, "Erc20Instance15Balances");
+impl_prefix!(ApprovesPrefix16, Instance16, "Erc20Instance16Balances");
+
+/// Alias for the Balance type for the provided Runtime and Instance.
+pub type BalanceOf<Runtime, Instance = ()> =
+	<Runtime as pallet_balances::Config<Instance>>::Balance;
+
+/// Storage type used to store approvals, since `pallet_balances` doesn't
+/// handle this behavior.
+/// (Owner => Allowed => Amount)
+pub type ApprovesStorage<Runtime, Instance> = StorageDoubleMap<
+	<Instance as InstanceToPrefix>::ApprovesPrefix,
+	Blake2_128Concat,
+	<Runtime as frame_system::Config>::AccountId,
+	Blake2_128Concat,
+	<Runtime as frame_system::Config>::AccountId,
+	<Runtime as pallet_balances::Config<Instance>>::Balance,
+>;
+
+#[precompile_utils::generate_function_selector]
+#[derive(Debug, PartialEq, num_enum::TryFromPrimitive, num_enum::IntoPrimitive)]
+pub enum Action {
+	TotalSupply = "totalSupply()",
+	BalanceOf = "balanceOf(address)",
+	Allowance = "allowance(address,address)",
+	Transfer = "transfer(address,uint256)",
+	Approve = "approve(address,uint256)",
+	TransferFrom = "transferFrom(address,address,uint256)",
+}
```

Comment:
Not directly related to this PR, but I wonder if the precompile utils should re-export parts of `num_enum` so we don't have to worry about syncing versions. Would it be possible for the `generate_function_selector` macro to derive those `TryFromPrimitive` and `IntoPrimitive` traits itself?